text/0000-lifecycle-all.md

```diff
@@ -0,0 +1,109 @@
+| `-buildpacks`   | optional | `CNB_BUILDPACKS_DIR`   | `/cnb/buildpacks`         | path to buildpacks directory |
+| `-daemon`       | optional | `CNB_USE_DAEMON`       | false                     | export to docker daemon |
+| `-uid`          | required | `CNB_USER_ID`          | -                         | UID of user in the stack's build and run images |
+| `-gid`          | required | `CNB_GROUP_ID`         | -                         | GID of user's group in the stack's build and run images |
+| `-version`      | optional | -                      | false                     | show version |
+| `-skip-restore` | optional | -                      | false                     | do not restore metadata from previous image or data from cache
+| `-tag` (multiples allowed) | optional | -           |                           | additional tags to apply to exported image
+
+Most of the these flags are existing lifecycle phase flags. `-skip-restore` is a a new addition that modifies the behavior of `analyzer` similarly to the existing `-skip-layers` flag, and, in addition, skips the `restorer` phase all together.
+
+This RFC proposes a `-tag` flag as the interface to provide additional tags, rather than multiple positional arguments (as is currently accepted by `exporter` ) for clarity and it's similarity to `docker build`.
+
+Some existing lifecycle flags (e.g. `-group` on the `detector` `builder` and `exporter` phases) are not necessary. They refer to files that were previously used to pass information between lifecycle phase processes. When the phases run in the same process, those values can be passed in memory.
+
+
+### User
+
+Right now, when `pack` runs w/o the `--publish` flag `analyzer` and `exporter` are run as root. This is required so that these phases can connect to the mounted docker daemon socket. Historically the `builder` and `detector` binaries are never run as root.
+
+In the daemon case pack will run `all` as `root` but the buildpacks' `/bin/detect` and `/bin/builder` scripts will be invoked as the provided user. This will require a change to build and detect implementations.
+
+### Credential Management
+
+Right now, when `pack` runs w/ the `--publish` flag `analyzer` and `exporter` are provided with registry credentials via the `CNB_REGISTRY_AUTH` environment variable.
+
+In the registry case pack will set `CNB_REGISTRY_AUTH` when invoking `all`. To prevents buildpacks from having read access to those credentials the build and detect implementations will ensure this variable is not present in the environment of the `/bin/detect` and `/bin/build` processes.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Making it easy to run all of the lifecycle in a single container will encourage folks to do so. However, this requires consumers to be more thoughtful about security and credential isolation. We will need to harden `builder` and `detector` implementation to ensure buildpacks are not given root access or registry credentials.
```

Comment:
Do we believe that it won't be possible to harden these implementations such that they can be trusted?  Do we think that no matter what we'll always need a fallback to multiple containers?

/cc @sclevine

Reply:
If we are running the lifecycle binaries that come packaged in the builder image, we can never harden them sufficiently to trust them. This is not because of potential problems with our implementation but rather, there is no guarantee that the files at those paths contain our implementation of the lifecycle.

We could append a layer to the image and essentially inject our lifecycle every build but this seems risky. There are numerous attack vectors if you don't trust the provenance of the builder image.